nova/virt/vmwareapi/images.py

```diff
@@ -399,6 +403,15 @@ def fetch_image_stream_optimized(context, instance, session, vm_name,
     LOG.info("Downloaded image file data %(image_ref)s",
              {'image_ref': instance.image_ref}, instance=instance)
     vmdk = vm_util.get_vmdk_info(session, imported_vm_ref)
+    if vmdk.capacity_in_bytes == INVALID_VMDK_SIZE:
+        LOG.warning("Deleting invalid template VM %s",
+                    vutil.get_moref_value(imported_vm_ref))
+        destroy_task = session._call_method(
+            session.vim,
+            "Destroy_Task",
+            imported_vm_ref)
+        session._wait_for_task(destroy_task)
```

Comment:
This could fail with another exception, because sometimes our template-vms end up without disk. I think we have to unregister them then. I did try to build something some years ago it seems:
```
    When deleting expired image templates, we sometimes fail, because the
    template is missing its disk. We don't know the reason, but we still
    want the template to be deleted and thus handle the InvalidVmState error
    now. In the failed deletion task of one of such templates in the
    vCenter, there's a longer message that suggests to mark the template as
    a VM, because the disk is missing. Therefore, we implement that behavior
    now and try to delete the VM again afterwards.
```

I'll write it down as a follow-up.